File: packages/polaris-sdk/src/transport.ts
import { PolarisError, PolarisTransportError } from "./errors";
import type { RateLimit, SDKResponse } from "./types";

type Fetch = typeof globalThis.fetch;

export type TransportOptions = {
  apiKey: string;
  baseUrl: string;
  fetch?: Fetch;
  maxRetries?: number;
};

type JSONRequest =
  | { method: "GET"; path: string; signal: AbortSignal | undefined }
  | {
      method: "POST" | "PATCH" | "DELETE";
      path: string;
      body?: unknown;
      idempotencyKey?: string;
      signal: AbortSignal | undefined;
    };

export class PolarisTransport {
  readonly apiKey: string;
  readonly baseUrl: string;
  readonly fetch: Fetch;
  readonly maxRetries: number;

  constructor(options: TransportOptions) {
    this.apiKey = validateAPIKey(options.apiKey);
    this.baseUrl = validateBaseURL(options.baseUrl);
    this.fetch = options.fetch ?? globalThis.fetch;
    if (typeof this.fetch !== "function") {
      throw new TypeError("Polaris requires a Fetch API implementation.");
    }
    this.maxRetries = normalizeMaxRetries(options.maxRetries);
  }

  url(path: string): string {
    return `${this.baseUrl}${path}`;
  }

  authorizationHeaders(): Headers {
    return new Headers({ Authorization: `Bearer ${this.apiKey}` });
  }

  async json<T>(request: JSONRequest): Promise<SDKResponse<T>> {
    let lastNetworkError: unknown;
    for (let attempt = 0; attempt <= this.maxRetries; attempt += 1) {
      try {
        const headers = this.authorizationHeaders();
        headers.set("Accept", "application/json");
        if ("body" in request && request.body !== undefined)
          headers.set("Content-Type", "application/json");
        if ("idempotencyKey" in request && request.idempotencyKey !== undefined)
          headers.set("Idempotency-Key", request.idempotencyKey);
        const response = await this.fetch(this.url(request.path), {
          method: request.method,
          headers,
          body:
            "body" in request && request.body !== undefined ? JSON.stringify(request.body) : null,
          signal: request.signal ?? null,
        });
        if (response.ok) {
          return {
            data: (response.status === 204 ? null : await response.json()) as T,
            ...responseMetadata(response),
          };
        }
        const error = await responseError(response);
        if (
          !isRetryableRequest(request) ||
          !isRetryableStatus(response.status) ||
          attempt === this.maxRetries
        ) {
          throw error;
        }
        await waitBeforeRetry(response, attempt, request.signal);
      } catch (error) {
        if (error instanceof PolarisError) throw error;
        if (request.signal?.aborted) throw error;
        lastNetworkError = error;
        if (!isRetryableRequest(request) || attempt === this.maxRetries) break;
        await delay(exponentialDelay(attempt), request.signal);
      }
    }
    throw new PolarisTransportError("Polaris request failed before a response was received.", {
      cause: lastNetworkError,
    });
  }
}

function isRetryableRequest(request: JSONRequest): boolean {
  return (
    request.method === "GET" ||
    ("idempotencyKey" in request && (request.idempotencyKey?.length ?? 0) > 0)
  );
}

export function responseMetadata(response: Response) {
  return {
    requestId: response.headers.get("X-Request-ID"),
    idempotencyReplayed: response.headers.get("Idempotency-Replayed") === "true",
    rateLimit: rateLimitFromHeaders(response.headers),
  };
}

export async function responseError(response: Response): Promise<PolarisError> {
  let envelope: unknown;
  try {
    envelope = await response.json();
  } catch {
    envelope = null;
  }
  const error = isRecord(envelope) && isRecord(envelope.error) ? envelope.error : null;
  const code = typeof error?.code === "string" ? error.code : "unexpected_response";
  const message =
    typeof error?.message === "string"
      ? error.message
      : `Polaris returned HTTP ${response.status}.`;
  const requestId =
    typeof error?.requestId === "string" ? error.requestId : response.headers.get("X-Request-ID");
  const details = isRecord(error?.details) ? error.details : null;
  return new PolarisError(
    response.status,
    code,
    message,
    requestId,
    details,
    positiveIntegerHeader(response.headers, "Retry-After"),
  );
}

function rateLimitFromHeaders(headers: Headers): RateLimit {
  return {
    limit: positiveIntegerHeader(headers, "RateLimit-Limit"),
    remaining: nonNegativeIntegerHeader(headers, "RateLimit-Remaining"),
    resetAfterSeconds: positiveIntegerHeader(headers, "RateLimit-Reset"),
  };
}

function positiveIntegerHeader(headers: Headers, name: string): number | null {
  const value = nonNegativeIntegerHeader(headers, name);
  return value !== null && value > 0 ? value : null;
}

function nonNegativeIntegerHeader(headers: Headers, name: string): number | null {
  const raw = headers.get(name);
  if (raw === null || !/^[0-9]+$/.test(raw)) return null;
  const value = Number(raw);
  return Number.isSafeInteger(value) ? value : null;
}

function validateAPIKey(value: string): string {
  const apiKey = value.trim();
  if (!apiKey.startsWith("syna_sa_") || apiKey.length <= "syna_sa_".length) {
    throw new TypeError("Polaris apiKey must be a syna_sa_ Service Account token.");
  }
  return apiKey;
}

function validateBaseURL(value: string): string {
  const parsed = new URL(value);
  if (
    (parsed.protocol !== "https:" && parsed.protocol !== "http:") ||
    parsed.username ||
    parsed.password
  ) {
    throw new TypeError("Polaris baseUrl must be an HTTP(S) URL without embedded credentials.");
  }
  if (parsed.search || parsed.hash) {
    throw new TypeError("Polaris baseUrl must not contain a query or fragment.");
  }
  return parsed.toString().replace(/\/+$/, "");
}

function normalizeMaxRetries(value: number | undefined): number {
  if (value === undefined) return 2;
  if (!Number.isSafeInteger(value) || value < 0 || value > 10) {
    throw new TypeError("Polaris maxRetries must be an integer between 0 and 10.");
  }
  return value;
}

function isRetryableStatus(status: number): boolean {
  return status === 429 || status === 500 || status === 502 || status === 503 || status === 504;
}

async function waitBeforeRetry(response: Response, attempt: number, signal?: AbortSignal) {
  const retryAfter = positiveIntegerHeader(response.headers, "Retry-After");
  await delay(retryAfter === null ? exponentialDelay(attempt) : retryAfter * 1000, signal);
}

function exponentialDelay(attempt: number): number {
  return Math.min(250 * 2 ** attempt, 4000);
}

export function delay(milliseconds: number, signal?: AbortSignal): Promise<void> {
  if (signal?.aborted) return Promise.reject(signal.reason);
  return new Promise((resolve, reject) => {
    const timeout = setTimeout(resolve, milliseconds);
    signal?.addEventListener(
      "abort",
      () => {
        clearTimeout(timeout);
        reject(signal.reason);
      },
      { once: true },
    );
  });
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

File: packages/polaris-python/src/polaris_agents/_transport.py
from __future__ import annotations

import json
import time
from dataclasses import dataclass
from typing import Any, Callable, Generic, Mapping, TypeVar, cast
from urllib.error import HTTPError, URLError
from urllib.parse import urlencode, urlsplit
from urllib.request import Request, urlopen

from ._errors import PolarisError, PolarisTransportError

T = TypeVar("T")
OpenURL = Callable[..., Any]


@dataclass(frozen=True)
class RateLimit:
    limit: int | None
    remaining: int | None
    reset_after_seconds: int | None


@dataclass(frozen=True)
class ResponseMetadata:
    request_id: str | None
    idempotency_replayed: bool
    rate_limit: RateLimit


@dataclass(frozen=True)
class SDKResponse(Generic[T]):
    data: T
    metadata: ResponseMetadata

    @property
    def request_id(self) -> str | None:
        return self.metadata.request_id

    @property
    def idempotency_replayed(self) -> bool:
        return self.metadata.idempotency_replayed

    @property
    def rate_limit(self) -> RateLimit:
        return self.metadata.rate_limit


class PolarisTransport:
    def __init__(
        self,
        *,
        api_key: str,
        base_url: str,
        max_retries: int = 2,
        timeout: float = 30.0,
        opener: OpenURL = urlopen,
        sleep: Callable[[float], None] = time.sleep,
    ) -> None:
        key = api_key.strip()
        if not key.startswith("syna_sa_") or len(key) <= len("syna_sa_"):
            raise ValueError("Polaris api_key must be a syna_sa_ Service Account token.")
        parsed = urlsplit(base_url)
        if parsed.scheme not in {"http", "https"} or not parsed.netloc or parsed.username or parsed.password:
            raise ValueError("Polaris base_url must be an HTTP(S) URL without embedded credentials.")
        if parsed.query or parsed.fragment:
            raise ValueError("Polaris base_url must not contain a query or fragment.")
        if not isinstance(max_retries, int) or isinstance(max_retries, bool) or not 0 <= max_retries <= 10:
            raise ValueError("Polaris max_retries must be an integer between 0 and 10.")
        if timeout <= 0 or timeout > 300:
            raise ValueError("Polaris timeout must be greater than 0 and at most 300 seconds.")
        self.api_key = key
        self.base_url = base_url.rstrip("/")
        self.max_retries = max_retries
        self.timeout = timeout
        self._opener = opener
        self._sleep = sleep

    def request_json(
        self,
        method: str,
        path: str,
        *,
        body: Mapping[str, Any] | None = None,
        query: Mapping[str, str | int] | None = None,
        idempotency_key: str | None = None,
    ) -> SDKResponse[dict[str, Any]]:
        url = self._url(path, query)
        encoded = json.dumps(body, separators=(",", ":")).encode() if body is not None else None
        retryable = method == "GET" or bool(idempotency_key)
        last_error: BaseException | None = None
        for attempt in range(self.max_retries + 1):
            request = Request(url, data=encoded, method=method, headers=self._headers(idempotency_key))
            try:
                response = self._opener(request, timeout=self.timeout)
                try:
                    payload = {} if response.getcode() == 204 else json.loads(response.read())
                    if not isinstance(payload, dict):
                        raise PolarisTransportError("Polaris returned a non-object JSON response.")
                    return SDKResponse(cast(dict[str, Any], payload), _metadata(response.headers))
                finally:
                    response.close()
            except HTTPError as error:
                parsed = _http_error(error)
                if not retryable or error.code not in {429, 500, 502, 503, 504} or attempt == self.max_retries:
                    raise parsed from error
                self._sleep(float(parsed.retry_after_seconds or _backoff(attempt)))
            except (URLError, OSError, TimeoutError) as error:
                last_error = error
                if not retryable or attempt == self.max_retries:
                    break
                self._sleep(float(_backoff(attempt)))
        raise PolarisTransportError("Polaris request failed before a response was received.") from last_error

    def open_event_stream(self, path: str, *, after_sequence: int) -> Any:
        request = Request(
            self._url(path, {"afterSequence": after_sequence}),
            method="GET",
            headers={"Authorization": f"Bearer {self.api_key}", "Accept": "text/event-stream"},
        )
        try:
            return self._opener(request, timeout=self.timeout)
        except HTTPError as error:
            raise _http_error(error) from error
        except (URLError, OSError, TimeoutError) as error:
            raise PolarisTransportError("Polaris Event stream failed before a response was received.") from error

    def _headers(self, idempotency_key: str | None) -> dict[str, str]:
        headers = {"Authorization": f"Bearer {self.api_key}", "Accept": "application/json"}
        if idempotency_key is not None:
            headers.update({"Content-Type": "application/json", "Idempotency-Key": idempotency_key})
        return headers

    def _url(self, path: str, query: Mapping[str, str | int] | None = None) -> str:
        suffix = f"?{urlencode(query)}" if query else ""
        return f"{self.base_url}{path}{suffix}"


def _metadata(headers: Any) -> ResponseMetadata:
    return ResponseMetadata(
        request_id=headers.get("X-Request-ID"),
        idempotency_replayed=headers.get("Idempotency-Replayed") == "true",
        rate_limit=RateLimit(
            limit=_integer_header(headers, "RateLimit-Limit", positive=True),
            remaining=_integer_header(headers, "RateLimit-Remaining", positive=False),
            reset_after_seconds=_integer_header(headers, "RateLimit-Reset", positive=True),
        ),
    )


def _http_error(error: HTTPError) -> PolarisError:
    try:
        envelope = json.loads(error.read())
    except (json.JSONDecodeError, UnicodeDecodeError):
        envelope = None
    value = envelope.get("error") if isinstance(envelope, dict) else None
    value = value if isinstance(value, dict) else {}
    details = value.get("details")
    return PolarisError(
        error.code,
        value.get("code") if isinstance(value.get("code"), str) else "unexpected_response",
        value.get("message") if isinstance(value.get("message"), str) else f"Polaris returned HTTP {error.code}.",
        value.get("requestId") if isinstance(value.get("requestId"), str) else error.headers.get("X-Request-ID"),
        details if isinstance(details, dict) else None,
        _integer_header(error.headers, "Retry-After", positive=True),
    )


def _integer_header(headers: Any, name: str, *, positive: bool) -> int | None:
    raw = headers.get(name)
    if not isinstance(raw, str) or not raw.isascii() or not raw.isdecimal():
        return None
    value = int(raw)
    if value < (1 if positive else 0):
        return None
    return value


def _backoff(attempt: int) -> float:
    return min(0.25 * (2**attempt), 4.0)
